### DISCUSSION TITLE
Empty response from request

### DISCUSSION THREAD

**Discussion by gee4vee** [octokit/octokit.js#2271] [Q&A]
I am trying to download a file from a repo in a non-async function. The `request` response is coming back as `{}`. What are the possible causes for getting back an empty response object?

```typescript
      let octokit: Octokit;
      const authToken = this.ghTokens.get(module.source.gitRepoUrl);
      if (authToken) {
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl(),
          auth: authToken
        });
      } else {
        logger.warn(`No authentication token found for repository ${module.source.gitRepoUrl}`);
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl()
        });
      }
      const response = (async () => {
        await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
          owner: module.source.getRepoOwner(),
          repo: module.source.getRepo(),
          path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
          ref: module.source.getBranch()
        });
      })();
```

**Comment by timrogers** [comment]
Can you try this, and tell me what output you get when you run it?

```js
let octokit: Octokit;
const authToken = this.ghTokens.get(module.source.gitRepoUrl);
if (authToken) {
  octokit = new Octokit({
    baseUrl: module.source.getGitHubUrl(),
    auth: authToken,
    log: console
  });
} else {
  logger.warn(`No authentication token found for repository ${module.source.gitRepoUrl}`);
  octokit = new Octokit({
    baseUrl: module.source.getGitHubUrl(),
    log: console
  });
}

(async () => {
  const response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
    owner: module.source.getRepoOwner(),
    repo: module.source.getRepo(),
    path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
    ref: module.source.getBranch()
  });

  console.log(response);
})();
```

It adds logging to the `Octokit` instances and property awaits `octokit.request` - at the moment, you are awaiting it inside your immediately involved function expresssion [IIFE], but then you aren't awaiting the IIFE itself.

**Comment by gee4vee** [comment]
@timrogers Thx for the tip! I'm trying your change but getting no output. I can see `console.log` output coming from statements outside of the IIFE, just not inside.

I tried grabbing the response outside the IIFE afterwards, but it's undefined.

```typescript
      let response: unknown;
      (async () => {
        response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
          owner: module.source.getRepoOwner(),
          repo: module.source.getRepo(),
          path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
          ref: module.source.getBranch()
        });
        // eslint-disable-next-line no-console
        console.log(`CONSOLE LOG GET response: ${response}`);
      })();
      logger.trace(`GET repo response: ${JSON.stringify(response)}`);
```

**Reply by timrogers** [reply]
Could you share the full code again? I'd like to try it.

**Reply by timrogers** [reply]
I am not surprised that you get a log from this line:

```
logger.trace(`GET repo response: ${JSON.stringify(response)}`);
```

...but `response` will be undefined because you haven't awaited the result of the IIFE.

**Reply by gee4vee** [reply]
If I add `await` before the IIFE, that last `logger.trace` line never gets logged.

**Reply by timrogers** [reply]
I tried the code locally and it does work for me. I get `GET repo response: undefined`, but I also get `CONSOLE LOG GET response: [object Object]`.

 In your case, how long does it run for before it exits? What Node version are you running?

**Reply by gee4vee** [reply]
I'm on v16.14.0. The whole test suite runs in about 6 seconds. Perhaps it has to do with Jest? I am using `--runInBand`.

**Reply by timrogers** [reply]
Are you able to run the code successfully on your local machine outside of the Jest context? That would be a good thing to check, as the code (with some necessary changes) does work for me.

**Reply by gee4vee** [reply]
I'll try a simple program. Will it be possible to get a response outside the IIFE, or do I need to put all code using `response` inside?

**Reply by gee4vee** [reply]
Of course, if I can't get it to work within Jest, that's a problem since I can't automate the testing :)

**Reply by timrogers** [reply]
It should be possible to use the response outside, but you would have to either use `await` or `.then`, because otherwise the following lines will run before the Octokit call finishes.

**Reply by gee4vee** [reply]
If I use `await` at the front of the IIFE, the `logger.trace` call doesn't produce any output. Does that mean it just never finishes executing in time? Does that mean it's not really waiting for it to finish?

**Reply by timrogers** [reply]
That sounds likely to me. There could be a networking issue with reaching your GitHub API - especially if it isn't `api.github.com`.

**Reply by gee4vee** [reply]
So is it actually possible to use OctoKit in a non-async context?

**Reply by timrogers** [reply]
I'm afraid that isn't possible because the HTTP calls are fundamentally asynchronous. You can either `await` the results or chain `.then`/`.catch` to handle the control flow. It should be possible to work with async code in Jest though - they have some docs on it.

**Comment by gee4vee** [comment]
```typescript
      // set up octokit
      let octokit: Octokit;
      const authToken = this.ghTokens.get(module.source.gitRepoUrl);
      if (authToken) {
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl(),
          auth: authToken,
          log: console
        });
      } else {
        logger.warn(`No authentication token found for repository ${module.source.gitRepoUrl}`);
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl(),
          log: console
        });
      }
      let response: unknown;
      (async () => {
        response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
          owner: module.source.getRepoOwner(),
          repo: module.source.getRepo(),
          path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
          ref: module.source.getBranch()
        });
        // eslint-disable-next-line no-console
        console.log(`CONSOLE LOG GET response: ${response}`);
      })();
      logger.trace(`GET repo response: ${JSON.stringify(response)}`);
```

**Comment by gee4vee** [comment]
@timrogers I reworked my stack to use async and await throughout. However when it gets to the `await octokit.request` call, it simply returns and continues execution. I can see it happen in the debugger. Does this usage look right to you?

```typescript
      // set up octokit
      let octokit: Octokit;
      const authToken = this.ghTokens.get(module.source.gitRepoUrl);
      if (authToken) {
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl(),
          auth: authToken,
          log: console
        });
      } else {
        logger.warn(`No authentication token found for repository ${module.source.gitRepoUrl}`);
        octokit = new Octokit({
          baseUrl: module.source.getGitHubUrl(),
          log: console
        });
      }
      // this is to avoid making parseModule async.
      logger.debug(
        `Retrieving metadata file from repo ${module.source.getRepoUrl()} in path ${
          module.source.gitFolder
        } in branch ${module.source.getBranch()}`
      );
      try {
        const response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
          owner: module.source.getRepoOwner(),
          repo: module.source.getRepo(),
          path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
          ref: module.source.getBranch()
        });
        logger.trace(`GET repo response: ${JSON.stringify(response)}`);
      } catch (error) {
        logger.error(`Error getting metadata file from repository ${module.source.getRepoUrl()}: ${error}`);
      }
```

**Reply by timrogers** [reply]
I just tried that code locally with Node v16.14.2 and I see the `GET repo response` log line as expected, with the details on the file I requested.

Can you try pointing to `README.md` in `octokit/octokit.js`, and see if you have issues with that?

**Reply by gee4vee** [reply]
Looks like same behavior with this (also changed the baseUrl to github.com)

```
        const response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
          // owner: module.source.getRepoOwner(),
          // repo: module.source.getRepo(),
          // path: `${module.source.gitFolder ? `${module.source.gitFolder}/` : ''}${METADATA_FILE_NAME}`,
          // ref: module.source.getBranch()
          owner: 'octokit',
          repo: 'octokit.js',
          path: 'README.md'
        });
```

All the functions in the stack are `async`, including the Jest tests, so I think all that should be correct.

**Reply by timrogers** [reply]
Let's strip this down and go for the simplest possible example! Does the following JavaScript snippet run for you on its own, if you set your token in the `GH_TOKEN` environment varibale?:

```js
// Replace me with `require` if you're using CommonJS modules 
import { Octokit } from 'octokit';

const octokit = new Octokit({
  auth: process.env.GH_TOKEN
});

(async () => {
  const response = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
    owner: 'octokit',
    repo: 'octokit.js',
    path: 'README.md',
    ref: 'main'
  });

  console.log(response)
})();
```

**Reply by gee4vee** [reply]
That sample worked, even without setting `GH_TOKEN`.

**Reply by timrogers** [reply]
That makes sense - the repo is available publicly 😊 Now I'd recommend gradually working from that sample to your code and trying to figure out what part breaks it, if you can. I can't see anything obvious.

**Reply by gee4vee** [reply]
@timrogers I figured it out... I was missing an await keyword in the test code :( Thanks for all your help!
